### PR TITLE
Add Oracle connection pool

### PR DESCRIPTION
**Pull request by eskabetxe**
as asked here: https://github.com/prestosql/presto/pull/1959#issuecomment-605538917

this is to allow to configure a oracle connection pool

### CODE REVIEW

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConnectorFactory.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        implements ConnectionFactory
+{
+    private final String connectionUrl;
+    private final boolean connectionPoolEnabled;
+    private final int connectionPoolMinSize;
+    private final int connectionPoolMaxSize;
+    private final Properties connectionProperties;
+    private final CredentialProvider credentialProvider;
+
+    private PoolDataSource pds;
```
Comment (Praveen2112) [inline presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConnectorFactory.java:44]: Can we have an interface for `ConnectionProvider` and have implementation for for both pooledDataSource and a driver based one and bind them in the run time ?

**Comment by tooptoop4** [comment]
@eskabetxe can u change 

```private final int fetchSize = 1000;```
to 10000 or at least 5000?

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConnectorFactory.java (hunk 1 above)
Comment (eskabetxe) [reply presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConnectorFactory.java]: hi @Praveen2112, I change this factory to be only a pool factory..
for driver based I leave the DriverConnectorFactory that is already used..
this was a old code from previous PR.

if you could check this changes.

**Comment by eskabetxe** [comment]
@tooptoop4 I create a PR with your [request](https://github.com/prestosql/presto/pull/3775)

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolDistributedQueries.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.oracle;
+
+import io.prestosql.testing.QueryRunner;
+import io.prestosql.tpch.TpchTable;
+
+public class TestOraclePoolDistributedQueries
```
Comment (losipiuk) [inline presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolDistributedQueries.java:19]: IMO running whole `DistributedQueries` is superfluous. It adds a lot of test runtime. `TestOraclePoolIntegrationSmokeTest` should be enough.

File: presto-oracle/pom.xml [hunk 3]
```diff
@@ -16,13 +16,20 @@
 
     <properties>
         <air.main.basedir>${project.parent.basedir}</air.main.basedir>
+        <dep.oracle.version>19.3.0.0</dep.oracle.version>
```
Comment (losipiuk) [inline presto-oracle/pom.xml:19]: Please extract separate commit `Extract a variable with Oracle JDBC version`

Comment (eskabetxe) [reply presto-oracle/pom.xml:19]: done

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolDistributedQueries.java (hunk 2 above)
Comment (eskabetxe) [reply presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolDistributedQueries.java]: done

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOracleConfig.java [hunk 4]
```diff
@@ -33,6 +33,9 @@ public void testDefaults()
                 .setVarcharMaxSize(4000)
                 .setTimestampDefaultPrecision(6)
                 .setNumberDefaultScale(10)
+                .setConnectionPoolEnabled(false)
```
Comment (losipiuk) [inline presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOracleConfig.java:36]: Is there a reason to have connection pooling disable by default? I would suggest having `true` here.

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOracleDistributedQueries.java [hunk 5]
```diff
@@ -47,9 +47,13 @@ protected QueryRunner createQueryRunner()
...
-        return OracleQueryRunner.createOracleQueryRunner(
-                oracleServer,
-                TpchTable.getTables());
+        return createOracleQueryRunner(oracleServer, TpchTable.getTables());
+    }
+
+    protected QueryRunner createOracleQueryRunner(TestingOracleServer server, Iterable<TpchTable<?>> tables)
+            throws Exception
+    {
+        return OracleQueryRunner.createOracleQueryRunner(server, tables);
```
Comment (losipiuk) [inline presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOracleDistributedQueries.java:56]: Inline this back it is no longer needed .

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.oracle;
+
+import io.prestosql.testing.QueryRunner;
+import io.prestosql.tpch.TpchTable;
+
+public class TestOraclePoolIntegrationSmokeTest
+        extends TestOracleIntegrationSmokeTest
```
Comment (losipiuk) [inline presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java:20]: Please extract abstract `BaseOracleIntegrationSmokeTest` with all the test changes needed for Oracle, but with `createQueryRunner` not implemented.
And then derive `TestOraclePoolIntegrationSmokeTest` and `TestOracleIntegrationSmokeTest` from it.
With that we do not extend non-abstract class. And we can get rid of `createOracleQueryRunner`.

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOracleConfig.java (hunk 4 above)
Comment (eskabetxe) [reply presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOracleConfig.java]: none.. changed

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOracleDistributedQueries.java (hunk 5 above)
Comment (eskabetxe) [reply presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOracleDistributedQueries.java]: done

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java (hunk 6 above)
Comment (eskabetxe) [reply presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java]: @losipiuk I left the createOracleQueryRunner, so initialization of server and table definition could be on BaseTest as this are equal to both implementations

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    public OraclePoolConnectorFactory(
+            String connectionUrl,
+            Properties connectionProperties,
+            CredentialProvider credentialProvider,
+            int connectionPoolMinSize,
+            int connectionPoolMaxSize)
+            throws SQLException
+    {
+        this.credentialProvider = credentialProvider;
+        this.connectionPool = PoolDataSourceFactory.getPoolDataSource();
```
Comment (losipiuk) [inline presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java:44]: rename `connectionPool` to `dataSource`

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+        //Setting connection properties of the data source
+        this.connectionPool.setConnectionFactoryClassName("oracle.jdbc.pool.OracleDataSource");
+        this.connectionPool.setURL(connectionUrl);
+
+        //Setting pool properties
+        this.connectionPool.setInitialPoolSize(connectionPoolMinSize);
+        this.connectionPool.setMinPoolSize(connectionPoolMinSize);
+        this.connectionPool.setMaxPoolSize(connectionPoolMaxSize);
+        this.connectionPool.setConnectionProperties(connectionProperties);
```
Comment (losipiuk) [inline presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java:54]: I would suggest adding `connectionPool.setValidateConnectionOnBorrow(true);`

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java (hunk 7 above)
Comment (eskabetxe) [reply presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java]: done

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java (hunk 8 above)
Comment (eskabetxe) [reply presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java]: done

File: presto-oracle/pom.xml [hunk 9]
```diff
@@ -16,13 +16,20 @@
...
     <dependencies>
         <dependency>
             <groupId>com.oracle.ojdbc</groupId>
             <artifactId>ojdbc8</artifactId>
-            <version>19.3.0.0</version>
+            <version>${dep.oracle.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId> com.oracle.ojdbc</groupId>
```
Comment (electrum) [inline presto-oracle/pom.xml:30]: Nit: remove leading space

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleClientModule.java [hunk 10]
```diff
@@ -51,10 +51,20 @@ public static ConnectionFactory connectionFactory(BaseJdbcConfig config, Credent
...
-                credentialProvider);
+        if (oracleConfig.isConnectionPoolEnabled()) {
+            return new OraclePoolConnectorFactory(
+                    config.getConnectionUrl(),
+                    connectionProperties,
+                    credentialProvider,
+                    oracleConfig.getConnectionPoolMinSize(),
+                    oracleConfig.getConnectionPoolMaxSize());
+        }
+        else {
```
Comment (electrum) [inline presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleClientModule.java:62]: Nit: the "else" is redundant since the above always returns

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java [hunk 11]
```diff
@@ -95,4 +98,43 @@ public OracleConfig setTimestampDefaultPrecision(int timestampDefaultPrecision)
...
+    }
+
+    @Config("oracle.connection-pool.enabled")
+    public OracleConfig setConnectionPoolEnabled(boolean connectionPoolEnabled)
+    {
+        this.connectionPoolEnabled = connectionPoolEnabled;
+        return this;
+    }
+
+    @Min(1)
```
Comment (electrum) [inline presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java:115]: Can this be 0?

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java [hunk 12]
```diff
@@ -95,4 +98,43 @@ public OracleConfig setTimestampDefaultPrecision(int timestampDefaultPrecision)
...
+        return connectionPoolMaxSize;
+    }
+
+    @Config("oracle.connection-pool.max-size")
+    public OracleConfig setConnectionPoolMaxSize(int connectionPoolMaxSize)
+    {
+        this.connectionPoolMaxSize = connectionPoolMaxSize;
+        return this;
+    }
 }
```
Comment (electrum) [inline presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java:140]: Add a validation:
```java
@AssertTrue(message = "Pool min size cannot be larger than max size")
public boolean isPoolSizedProperly()
{
    return getMaxPoolSize() >= getMinPoolSize();
}

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+            CredentialProvider credentialProvider,
+            int connectionPoolMinSize,
+            int connectionPoolMaxSize)
+            throws SQLException
+    {
+        this.credentialProvider = credentialProvider;
+        this.connectionPool = PoolDataSourceFactory.getPoolDataSource();
+
+        //Setting connection properties of the data source
+        this.connectionPool.setConnectionFactoryClassName("oracle.jdbc.pool.OracleDataSource");
```
Comment (electrum) [inline presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java]: Can we use `OracleDataSource.class.getName()` here?

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java (hunk 11 above)
Comment (eskabetxe) [reply presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java:115]: it shouldnt.. you will not have any connection

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java (hunk 13 above)
Comment (eskabetxe) [reply presto-oracle/src/main/java/io/prestosql/plugin/oracle/OraclePoolConnectorFactory.java]: done

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java (hunk 12 above)
Comment (eskabetxe) [reply presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java:140]: done

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleClientModule.java (hunk 10 above)
Comment (eskabetxe) [reply presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleClientModule.java:62]: done

File: presto-oracle/pom.xml (hunk 9 above)
Comment (eskabetxe) [reply presto-oracle/pom.xml:30]: done

File: presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java [hunk 14]
```diff
@@ -67,4 +71,49 @@ public OracleConfig setNumberRoundingMode(RoundingMode numberRoundingMode)
...
+    }
+
+    @Config("oracle.connection-pool.enabled")
+    public OracleConfig setConnectionPoolEnabled(boolean connectionPoolEnabled)
+    {
+        this.connectionPoolEnabled = connectionPoolEnabled;
+        return this;
+    }
+
+    @Min(0)
```
Comment (losipiuk) [inline presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java:88]: nit: add tests for validation constraints

Comment (eskabetxe) [reply presto-oracle/src/main/java/io/prestosql/plugin/oracle/OracleConfig.java:88]: done

**Comment by eskabetxe** [comment]
@losipiuk could this be merged?

**Comment by eskabetxe** [comment]
@Praveen2112 done

**Comment by losipiuk** [comment]
Checkstyle failed.
I think what @Praveen2112 asked for was to separate integration tests refactoring (extracting base class) from adding tests for connection pooling. It would be a bit nicer for review. Though given fact review is done already I will not push for that.

**Comment by eskabetxe** [comment]
@losipiuk fixed
@Praveen2112 I generate a commit with the extraction of base class..

File: presto-oracle/pom.xml [hunk 15]
```diff
@@ -14,13 +14,20 @@
...
+            <artifactId>ucp</artifactId>
```
Comment (findepi) [inline presto-oracle/pom.xml:29]: This is being added in `Extract a variable with Oracle JDBC version` commit. Does not look intentional.

(Not sure why would you want to separate this btw, but maybe this was agreed upon)

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java [hunk 16]
```diff
@@ -13,82 +13,16 @@
...
-
-import static io.prestosql.spi.type.VarcharType.VARCHAR;
-import static io.prestosql.testing.assertions.Assert.assertEquals;
-import static io.prestosql.tpch.TpchTable.*;
-import static io.prestosql.tpch.TpchTable.REGION;
-import static org.assertj.core.api.Assertions.assertThat;
 
 public class TestOraclePoolIntegrationSmokeTest
-        extends AbstractTestIntegrationSmokeTest
+        extends BaseOracleIntegrationSmokeTest
```
Comment (findepi) [inline presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java:20]: the `Oracle IntegrationTest cleanups` should not be a separate commit
(we should not add `TestOraclePoolIntegrationSmokeTest` being a copy of TestOracleIntegrationSmokeTest` and only then  extract common super class)

**Comment by findepi** [comment]
(just skimming)

File: presto-oracle/pom.xml (hunk 15 above)
Comment (eskabetxe) [reply presto-oracle/pom.xml:29]: was asked here https://github.com/prestosql/presto/pull/3770#discussion_r429880053

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java (hunk 16 above)
Comment (eskabetxe) [reply presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java:20]: should I squash the 3 commits?

File: presto-oracle/pom.xml (hunk 15 above)
Comment (losipiuk) [reply presto-oracle/pom.xml:29]: I think that at the time i suggested extracting the commit there was just one use of `dep.oracle.version`. But maybe I missed the other one.

Anyway. 
* Either leave commit which extract variable and only use it for `ojdbc8` entry. And then add `ucp` in followup commit together with pooling implementation.
* Or squash two commits together.

I am slightly toward former one.

File: presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java (hunk 16 above)
Comment (losipiuk) [reply presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java:20]: It should be either:

2 commits:
- rafactor tests extractin `BaseOracleIntegrationSmokeTest` (still with 1 concrete subclass `TestOracleIntegrationSmokeTest`)
- implement pooling and add `TestOraclePoolIntegrationSmokeTest `

Or just 1 commit which refactors tests, add pooling implementation and `TestOraclePoolIntegrationSmokeTest` alltogether.

Comment (Praveen2112) [reply presto-oracle/src/test/java/io/prestosql/plugin/oracle/TestOraclePoolIntegrationSmokeTest.java:20]: I guess the `OracleQueryRunner` changes can be treated as a separate commit while other the tests can be a part of Add Oracle connection pool

**Comment by losipiuk** [comment]
@eskabetxe there are two minor comments still. 
 - one about variable extraction in pom.xml: https://github.com/prestosql/presto/pull/3770#discussion_r440106555
 - the other one about splitting tests refactoring around commits: https://github.com/prestosql/presto/pull/3770#discussion_r440107366

Are you planning to address those? It would be nice to merge this one eventually :)

**Comment by eskabetxe** [comment]
@losipiuk sorry for the delay.. 
its done...

**Comment by losipiuk** [comment]
Thanks. Looks good modulo checkstyle violation:
```
2020-07-06T19:18:59.5653900Z [INFO] There is 1 error reported by Checkstyle 8.32 with checkstyle/airbase-checks.xml ruleset.
2020-07-06T19:18:59.5678894Z [ERROR] src/test/java/io/prestosql/plugin/oracle/OracleQueryRunner.java:[16,8] (imports) UnusedImports: Unused import - com.google.common.collect.ImmutableList.
```

**Comment by eskabetxe** [comment]
@losipiuk fixed

**Comment by losipiuk** [comment]
Merged, thanks!
